Improve linking inside guides

robin850 opened the pull request:
> Original proposal : https://gist.github.com/robin850/56848c86793ed7d5f0cafdf2b11d931a

Hi,

This pull request improves a bit linking to the API site or GitHub in the guides.

Actually, if `RAILS_VERSION` is specified generating the guides, it's automatically injected in `http://api.rubyonrails.org` URLs if there isn't a version already.

Also, now notes like "Defined in `path/to/file.rb`" are automatically expanded to a link so people can see the file on GitHub, tied to the version the guides were released against.

There's already an opened pull request to address this last point. The author has been properly credited inside the commit but here's the PR for future references: https://github.com/rails/rails/pull/20036.

Have a nice day ! :-)

robin850 committed:
Automatically inject the current Rails version in API links [ci skip]

To make sure that the user won't look at a feature that doesn't
already exist if they are looking at a previous version of the guides,
let's automatically inject the Rails version the guides have been
generated against.

matthewd commented:
This doesn't help with people looking at the raw files on GitHub, for example.

Perhaps instead of using `api://`, we could match `https://api.rubyonrails.org/` -- so, the source links always point to the main/current version, and the compile step just updates them to edgeapi / the versioned path as appropriate?

As a bonus challenge, ISTM we'd be better keeping people in the version they arrived in. So even though they're looking at the docs for v5.0.0.1 at `/` right now, we don't want to bounce them into `/v5.0.0.1/`. Achieving that seems like it would require either JS, or dynamic server-side evaluation, though. :confused:

sgrif commented:
Is it possible to have `github://` link to the commit it was built against instead of master when `RAILS_VERSION` is unset?

utilum commented:
Don't you guys open links from your text editor? Too bad to break that.

robin850 committed:
Automatically link to Ruby files referenced in notes [ci skip]

To ease reading the "Active Support Core Extensions" guide, let's
automatically link references to Ruby files.

It's also possible to reference other components' files in the even
though it's not used in the guides at the moment.

[Petr Skocik & Robin Dupret]

robin850 commented:
Hi there,

Sorry it took so long to get back to this ! Now, the renderer is injecting the version inside API URLs without relying on any `api://` scheme. Injecting the version inside GitHub URLs isn't done anymore since running `grep` on the guides folder shows that there are only links to pull requests or pages that aren't tied to any version.

> Is it possible to have `github://` link to the commit it was built against instead of master when `RAILS_VERSION` is unset?

For the reason explained above, I don't think this is necessary and whenever a guide would be saved, the commit would be erased. 😕

> So even though they're looking at the docs for v5.0.0.1 at / right now, we don't want to bounce them into `/v5.0.0.1/`.

I don't understand what you want to do here, could you explain it please ? :blush:

fxn commented:
@robin850 I believe @matthewd means that if you are browsing for example `/v5.0/`, it would be ideal to keep that in the links. This is from guides to API though, there is a context change, and as he points out it would need dynamic code somewhere. So it is a nice to have but I believe not needed for this PR.

There is no 5.0 generation by the way, those logical labels are symlinks in the docs server (see the files layout in the comment at the top of [this file](https://github.com/rails/rails-docs-server/blob/1e16e4891df82bc269b072597b8801dede9f879c/lib/docs_generator.rb)).

The proposal seems good to me, I only have feedback about code style. Would suggest writing this method

```ruby
def link(url, title, content)
  if url.start_with?('http://api.rubyonrails.org')
    %(<a href="#{api_link(url)}">#{content}</a>)
  elsif title
    %(<a href="#{url}" title="#{title}">#{content}</a>)
  else
    %(<a href="#{url}">#{content}</a>)
  end
end
```

as

```ruby
def link(url, title, content)
  if url.start_with?('http://api.rubyonrails.org')
    %(<a href="#{api_link(url)}">#{content}</a>)
  else
    super
  end
end
```

Then, regarding the regular expression

```ruby
/^\s*NOTE:\s+Defined\s+in\s+\<code\>([^`]*)\<\/code\>\.?$/
```

in Ruby regular expressions "<" and ">" are not metacharacters, special labels only work if they are at column 0 (as can be seen in the existing regexp), and the file paths the Markdown renderer receives have no backquotes. Would suggest this other one:

```ruby
%r{^NOTE:\s+Defined\s+in\s+<code>(.*?)</code>\.$/
```

On the other hand, this

```ruby
path = [root, rest].join('/')
```

is saying that `file_path` is the one we wanted after all. Maybe this could be made more concise.

Last, the method `api_link` modifies the argument in place. You normally do not want to do that, the caller should have its string untouched. And as a fine touch just return `url` without explicit `return`, to match the style in the rest of methods.

Could you have a look at this feedback?

matthewd commented:
> I believe @matthewd means that if you are browsing for example `/v5.0/`, it would be ideal to keep that in the links

Exactly. I agree it's fine to leave for another PR.

I think the way to do it would be to store the files on disk with the fully versioned path (as this PR does), then have on-load JS that rewrites all the links to match the [possibly truncated] version in the actual current URL.

robin850 commented:
Thanks for your review @fxn, most of your comments are addressed !

> I believe @matthewd means that if you are browsing for example /v5.0/, it would be ideal to keep that in the links.

Ah, yes, that'd be great !

> then have on-load JS that rewrites all the links to match the [possibly truncated] version in the actual current URL.

I don't know how if it's feasible or not but can't the API server look at the `HTTP_REFERER` header or something like that and do a redirect if it matches `/\d.\d/` ? Thus, the user wouldn't have to load an extra asset.

> Would suggest writing this method [with `super`]

Actually, that's something that Redcarpet doesn't support at the moment because, maybe in the future though. :blush:

fxn commented:
Hey, the leading `\s*` in the regexp is fighting for survival, all hope is futile! Let's remove it!

> I don't know how if it's feasible or not but can't the API server look at the `HTTP_REFERER` header or something like that and do a redirect if it matches `/\d.\d/` ? Thus, the user wouldn't have to load an extra asset.

Yes, I have sometimes wondered about such approach but have always come out with doubts. For example, the fact that a request to the same static public resource may yield a 200 or a 302 depending on the referer... feels weird to me. There are some other use cases that I don't remember now that made me to not go in that direction.

Loading a resource is no big deal on the other hand, and in any case a one-liner in the HEAD of the document is no big deal at all.

We may definitely explore refining this in other patches. 

> Actually, that's something that Redcarpet doesn't support at the moment because, maybe in the future though. :blush:

Ahh, I see https://github.com/vmg/redcarpet/issues/51, that @vmg guy!!! 😄

fxn commented:
@robin850 I messed up your last comment without noticing, sorry about that!

robin850 commented:
> Hey, the leading \s* in the regexp is fighting for survival, all hope is futile! Let's remove it!

Haha, this is fixed, it should be good now !

> I messed up your last comment without noticing, sorry about that!

Never mind ! :-)

fxn commented:
Awesome, thanks man!
